### ISSUE TITLE
Generated exe files don't work on arm64

### ISSUE THREAD

**Issue by hamarb123** [dotnet/runtime#63144]
### Description

When making a project that generates an exe file on AnyCPU (the default), the exe file should work on Windows on arm.

### Reproduction Steps

Make a .NET Project that generates an exe file (eg. wpf, winforms, console).
Compile the project.
Install the .NET 6.0.1 Runtime (or Desktop Runtime, which is what I'm using) on the arm machine.

### Expected behavior

Running the exe on the arm machine file runs the program as expected.

### Actual behavior

Run the .exe file on Windows on arm and note that nothing happens, except the console output:
```
Failed to load the dll from [C:\Program Files\dotnet\host\fxr\6.0.1\hostfxr.dll], HRESULT: 0x800700C1
The library hostfxr.dll was found, but loading it from C:\Program Files\dotnet\host\fxr\6.0.1\hostfxr.dll failed
  - Installing .NET prerequisites might help resolve this problem.
     https://go.microsoft.com/fwlink/?linkid=798306
```

### Regression?

I think so - I'm pretty sure in earlier rc versions of .NET 6.0 it was opening, but I had other issues with these versions, which seem to be fixed in the latest version.

### Known Workarounds

Run `dotnet <path-to-dll>.dll` instead. This isn't a proper workaround since it doesn't use icons and manifests etc.

### Configuration

.NET 6.0.1
Windows 11.0.22000.376 arm64 (on Parallels).
It is specific to Windows on arm.

**Comment by tomrus88** [comment]
Pretty sure you must publish as win-arm64 in order for it to work.
https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#windows-rids

**Comment by hamarb123** [comment]
> Pretty sure you must publish your app as win-arm64 in order for it to work.
> https://docs.microsoft.com/en-us/dotnet/core/rid-catalog#windows-rids

What's the point of AnyCPU then? And I don't have to do it for x64 or x86 (both of these run as x64 and x86 processes respectively) and I didn't to have to for arm64 during the release candidates (I think).

If I publish it, can I publish it for all windows platforms with a shared executable, because that would be fine enough?

**Comment by hamarb123** [comment]
I'm going to publish per architecture :( - this increases my size by ~2x.

**Comment by huoyaoyuan** [comment]
> What's the point of AnyCPU then?

In .NET Core, there is no longer **AnyCPU exe**. The .NET Framework exe worked because Windows handles it. .NET Core was decoupled from Windows, so the approach no longer works.

> If I publish it, can I publish it for all windows platforms with a shared executable, because that would be fine enough?

Shared entry point is no longer possible, but standalone entry points (app.exe) + shared managed code (the real app.dll) should be possible, and I remember there's already proposal for it.

**Comment by hamarb123** [comment]
> In .NET Core, there is no longer **AnyCPU exe**. The .NET Framework exe worked because Windows handles it. .NET Core was decoupled from Windows, so the approach no longer works.

That's makes sense. But .exes are only relevant on Windows anyway in .NET Core, so I don't see why this couldn't be done. A similar thing could also be done for macOS with Universal Binaries.

> Shared entry point is no longer possible, but standalone entry points (app.exe) + shared managed code (the real app.dll) should be possible, and I remember there's already proposal for it.

Do you know where this proposal is?

Thanks

**Comment by tomrus88** [comment]
> That's makes sense. But .exes are only relevant on Windows anyway in .NET Core, so I don't see why this couldn't be done. A similar thing could also be done for macOS with Universal Binaries.

The thing is every platform has different format of executable file (PE/ELF/Macho etc), even Windows x64 and Windows ARM have different formats (machine code is different). This means you need separate executable file for every platform you want your app to work on.

Your app code is still cross platform, but executable can't be.

> Do you know where this proposal is?

Would like to know as well.

**Comment by hamarb123** [comment]
> This means you need separate executable file for every platform you want your app to work on.

I know, and it's unfortunate, but there's nothing we can do about that. But in my opinion, it's better to at least have an option to have an executable per platform than an executable per platform and architecture (where possible, likely just macOS and Windows). This would be especially relevant for something like Wpf and WinForms, or net6.0-macos where it  seems silly to have to pick a architecture in advance, especially since Apple recommends using Universal Binaries.

> Yes on Mac there's Universal binaries and you can make single executable that works on both Intel and ARM chips, but on Windows there's no such thing.

.NET Framework has it already, just not for arm yet since it doesn't support arm yet (but it used to work with Itanium, which is different to x64 and x86). 4.8.1 will support arm64 and I would be very surprised if this didn't work, I see no reason why the same mechanism couldn't be used with .NET Core.

**Comment by hamarb123** [comment]
> On original issue: technically if you build your app on Windows ARM device, generated executable should run fine on that device, because its being generated for current platform by default. So there may be indeed some bug...

I was building on seperate x64 machine, so probably no issue.

**Comment by tomrus88** [comment]
> .NET Framework has it already, just not for arm yet since it doesn't support arm yet (but it used to work with Itanium, which is different to x64 and x86). 4.8.1 will support arm64 and I would be very surprised if this didn't work, I see no reason why the same mechanism couldn't be used with .NET Core.

Its very different in .NET Framework, because it's deeply integrated into OS, while .NET Core is decoupled from OS.
.NET Framework exe isn't native exe, it's managed exe and OS uses integrated .NET Framework runtime on execute it.

For .NET Core you need native executable that will spawn .NET Core runtime and execute your app code on that runtime. This is because runtime no longer integrated into OS.

**Comment by hamarb123** [comment]
That's a shame. Perhaps something could be done about this in the future.

**Comment by vitek-karas** [comment]
If you don't need an executable for your app (for example it's a tool which you run through scripts) you can start the app via `dotnet app.dll` - in this case the "executable" is provided by the machine and thus should be the right architecture always. This is what's called "portable app" for .NET Core - and it's basically its version of AnyCPU (it's actually better as it can be any architecture and any platform). If you need an executable, you need one per target platform/arch. As mentioned above there can be some tricks to avoid all combinations (universal binaries on macOS, win-x86 works on both x86 and x64 and so on), but ultimately there's no good solution for this.

I'm not aware of a proposal to generate multiple executables like mentioned above (but it's very possible I forgot), so if somebody knows where it is, please link it here. Technically the SDK can generate an executable for any platform and with some work it should be possible to generate multiple such executables, but it's unclear how this should look from user's perspective:
* It would only save space for framework dependent apps (Self-contained apps obviously need platform specific framework in them)
* Where to write the multiple executables? Different file names, not everybody would want this? Different directories (but then it would not run out of the box, ...)?
* How to specify the list of RIDs one wants the executables fors? `RuntimeIdentifiers` might be the way, but it currently means different thing - so it would be a breaking change.

It would be good if you would describe how you would want this to work and how would you deliver such application to the target device. (to give us a better idea what are the requirements)

**Comment by hamarb123** [comment]
> I'm not aware of a proposal to generate multiple executables like mentioned above (but it's very possible I forgot)

> Shared entry point is no longer possible, but standalone entry points (app.exe) + shared managed code (the real app.dll) should be possible, and I remember there's already proposal for it.

I think they were saying that the dll should be the same for all of them, which it currently isn't - this is a big waste of space if a meaningful amount of logic is in here (ie. if I have a project that generates an app.exe & app.dll, when I build it on different architectures, the app.dll changes, but when I reference another project, its library.dll doesn't change - this is what happens today, and I think it shouldn't unless there's a good reason for app.dll to be different).

> How to specify the list of RIDs one wants the executables fors? RuntimeIdentifiers might be the way, but it currently means different thing - so it would be a breaking change.

I think the current method should remain the same, but perhaps we could make RID groups (which would only work in non-self contained mode, there could also probably be issues for natives with this method, perhaps these could just be put in folders since this mode would be more for the developer to do what they want than for immediate distribution)? ie. a new RID could be `win-x64,win-x86,win-arm64` which would represent all 3 RIDs in one RID, this would then generate `app-win-x64.exe`, etc. (there is likely a better way to do this). The point of this mode is that the developer could be sure that (other than the exes), everything is going to be exactly the same across all platforms, except the platform-specific folders mentioned earlier - this could probably be made today from just building each of those platforms and then applying a simple file comparison and then chucking in various folders, so perhaps adding this to .NET itself isn't necessary.

> It would be good if you would describe how you would want this to work and how would you deliver such application to the target device. (to give us a better idea what are the requirements)

There would be 2 ways that I will distribute it - one would be msix, which I assume would already work as expected. The other way is using my own distribution app, it is very flexible, so the most simple way would be the best ie. different .exe, but everything else is as same as can be, and I can just package this in the most logical way using a script.

Also, on a seperate note, https://docs.microsoft.com/en-us/dotnet/core/rid-catalog lists `win-arm` as a valid RID, I was under the impression that .NET Core only runs on arm64 on Windows, so what's happening here? Also, https://docs.microsoft.com/en-us/dotnet/core/install/windows?tabs=net60#dependencies lists some additional requirements for older platforms, are these automatically grabbed by the .NET Runtime installer?

**Comment by vitek-karas** [comment]
> app.dll changes between the targets

That's true, but in reality the difference typically doesn't matter. That is you can take the file built for arm64 and use it in x64 app and it will very likely work (I can't think of a case were it wouldn't, unless you have ifdefs around architectures in your code). I do agree that it would be nice to have a way in the SDK to force the file to be the same.

The reason why library.dll doesn't change is that it's not recompiled. Unfortunately the app.dll is always recompiled in this case.

> Producing multiple exes

Technically you should be able to do this (not supported, but it should basically always work):
* Build the app as "portable", this is the default if you don't specify any RID. If your app has native dependencies this is easy to recognize as the output would contain a `runtime` directory (which would contain platform specific copies of the native dependencies). Non-portable apps would never have that directory as the right one is picked an copied to the root folder.
* Build the app as RID specific but really only use the `.exe` from the build - throw everything else away. There are potentially cheaper ways to do this, but they require some SDK/msbuild magic which is tricky to maintain.

The "portable" app is built in such a way that it can run correctly on all targets through the `dotnet app.dll` invocation. Using the `app.exe` is basically exactly the same as running `dotnet app.dll`.

I do agree that if this is an important scenario it would be good to have direct support for this in the SDK.

Note that the question of specifying the inputs and how to produce the outputs still remains - we would have to find something which is possibly usable by everybody and which doesn't cause confusion.

> win-arm

The presence of the RID in the catalog doesn't mean that it's supported. It only means the runtime would recognize such target, but there's a long way from that to "supported" or even "works there". It's also possible that it should be removed to make the file smaller if it's not needed.

> Additional dependencies on down-level platforms

I don't think the installer will include these or "grab them". But I'm not an installer expert...

**Comment by hamarb123** [comment]
> The reason why library.dll doesn't change is that it's not recompiled. Unfortunately the app.dll is always recompiled in this case.

What I'm suggesting is simply build the dll for "AnyCPU" and similarly for platform (I'm not sure if this is changed from the tfm when built with publish, but I don't think it should be) and then use it on all of them - if there are ifdefs, the dll can still be branded as "AnyCPU" and just be different - this is the behaviour I would have assumed was already present. This would reduce distribution size by ~10-15% for me and I barely even have any logic in this dll (on a side note: is it so big because the compiled resx is in both the exe and dll since I'm using the icon from it?).

> I don't think the installer will include these or "grab them". But I'm not an installer expert...

I'm also unsure if msix packager would get them, but I would assume that it essentially runs dotnet publish per platform and then copy the files if you check each cpu architecture when exporting it in VS.

> The "portable" app is built in such a way that it can run correctly on all targets through the dotnet app.dll invocation. Using the app.exe is basically exactly the same as running dotnet app.dll.

My current solution would probably be to publish as non self contained per architecture, and then find the files that are the same and only have these once in some sort of "common" folder - the purpose of this would be to reduce size.

> Build the app as "portable", this is the default if you don't specify any RID. If your app has native dependencies this is easy to recognize as the output would contain a `runtime` directory (which would contain platform specific copies of the native dependencies). Non-portable apps would never have that directory as the right one is picked an copied to the root folder.

Is this `runtime` folder automatically included in native library resolution eg. using `dlopen` through P/Invoke?

> Note that the question of specifying the inputs and how to produce the outputs still remains - we would have to find something which is possibly usable by everybody and which doesn't cause confusion.

I'll think about it more and see if I can come up with a better way of doing it.

**Comment by vitek-karas** [comment]
> Is this runtime folder automatically included in native library resolution eg. using dlopen through P/Invoke?

Yes

**Comment by vitek-karas** [comment]
> on a side note: is it so big because the compiled resx is in both the exe and dll since I'm using the icon from it?

If you have Win32 native resources embedded in the dll, they will get copied to the exe (that's how the icon gets there). Unfortunately we don't remove them from the dll even though they're effectively useless there. This is NOT the case for managed embedded resources though - those are only in the dll.

**Comment by jimm98y** [comment]
I had a similar requirement: we have a large app with many DLLs that is compiled for Any CPU and as we migrated it from NET Framework to NET Core, we wanted to add support for ARM64. Publishing it for x64 and ARM64 would have doubled our build time and also the output size. That's why I tried to make the build to compile only once (Portable, Framework-dependent, Any CPU) and get all the EXEs into the output folder. Using "dotnet <app>" is not possible in our case, because we needed the shortcuts in the Start menu with icons, version, copyrights, etc. My solution is here: [https://github.com/jimm98y/MultiTargetingSample](https://github.com/jimm98y/MultiTargetingSample)

It generates 3 EXEs in the publish output:
MultiTargetingSampleApp.exe (x64)
MultiTargetingSampleApp_x86.exe
MultiTargetingSampleApp_ARM64.exe

**Comment by agocke** [comment]
Closing as not planned. We can re-evaluate producing an ARM64X apphost if that becomes a more popular approach. It will unfortunately increase app size, as it will contain code for both architectures.

This deployment would likely be incompatible with R2R.
